taskflow/algorithm/for_each.hpp

```diff
@@ -75,36 +75,50 @@ TF_FORCE_INLINE auto make_for_each_task(B b, E e, C c, P part = P()) {
   };
 }
 
+template<typename T, typename C>
+using is_index_func = std::is_invocable_r<void, C, T>;
+
+template<typename T, typename C>
+using is_range_func = std::is_invocable_r<void, C, T, T>;
+
 // Function: make_for_each_index_task
-template <typename B, typename E, typename S, typename C, typename P = DefaultPartitioner>
-TF_FORCE_INLINE auto make_for_each_index_task(B b, E e, S s, C c, P part = P()){
+template <typename T, typename C, typename P = DefaultPartitioner>
+TF_FORCE_INLINE auto make_for_each_index_task(T b, T e, C c, P part = P()){
```

Comment:
The comment in the PR is specifically about this part of the patch. The begin and end iterators, `b` and `e` may sometimes be of different types and I think it is useful to retain the ability to have different types as was there before.

Reply:
Separately, though, I think it is important to keep the original interface for backward compatibility. Perhaps you want to overload the function by providing a second version that takes the two-argument function (instead of modifying the existing function)? You can use `std::enable_if_t` to ensure the right one gets called.